blockware/fireworks/src/main.cpp

```diff
@@ -11,73 +11,163 @@
+unsigned long lastWakeUpdate;
+unsigned long lastTextEnabled;
+unsigned long lastTextUpdate;
+unsigned int wakeMaxDuration = 90000;
 
 // NOTE: `loop`, `setup` and `flush` are sort of common operations
 // Modify `initialize` and `tick` to customize this demo
 
+void pickNextLogoColor() {
+  if (wakeupAll && (nextLogoColor <= 0 || sourceLogoColor <= 0)) {
+    nextLogoColor = std::max(sourceLogoColor, nextLogoColor);
+    sourceLogoColor = 0x0000;
+  }
+  
+  if (wakeupAll || nextLogoColor <= 0) {
+    while(sourceLogoColor == nextLogoColor || nextLogoColor <= 0) {
+      nextLogoColor = randomLogoColor();
+    }
+  } else {
+    nextLogoColor = 0x0000;
+  }
+}
+
+uint16_t getLogoColor() {
+  // When asleep, only keep color logo on screen for 2s, otherwise 10s
+  unsigned int onScreenTime = sourceLogoColor > 0 && !wakeupAll ? 2000 : 10000;
+  
+  // If we've just woken up, set logo color to animate in quickly
+  if (wakeupAll && (nextLogoColor == 0 || sourceLogoColor == 0)) {
+    uint16_t currStep = nextLogoColor == 0 ? (totalLogoColorSteps - logoColorStep) : logoColorStep;
+    logoColorStep = (currStep * fastLogoColorSteps / totalLogoColorSteps) + 1; // Add 1 to make sure we animate
+    totalLogoColorSteps = fastLogoColorSteps;
+    pickNextLogoColor();
+  }
+
+  if (logoColorStep >= totalLogoColorSteps) {
+    totalLogoColorSteps = defaultLogoColorSteps;
+    sourceLogoColor = nextLogoColor;
+    lastLogoColorUpdate = millis();
+    logoColorStep = 0;
+  } else if (logoColorStep > 0) {
+    // Update color to next step
+    logoColorStep += 1;
+    return getColorMix(sourceLogoColor, nextLogoColor, totalLogoColorSteps, logoColorStep);
+  } else if (millis() - lastLogoColorUpdate > onScreenTime) {
+    // It's time to start animating the color change
+    pickNextLogoColor();
+    logoColorStep = 1;
+  }
+
+  return sourceLogoColor;
+}
+
 void drawLogo()
 {
-    canvas->drawRGBBitmap(position.x, position.y, (uint16_t*)logo, tlWidth, tlHeight);
+  uint16_t logoColor = getLogoColor();
+  // Draw Logo 1 pixel at a time to allow changing the color
+  int16_t larryStartY = (SCREEN_HEIGHT - tlHeight) / 2;
+  int16_t larryStartX = (SCREEN_WIDTH - tlWidth) / 2;
+  for (uint8_t y = 0; y < tlHeight; y++) {
+    for (uint8_t x = 0; x < tlWidth; x++) {
+      uint16_t color = ((uint16_t *)tl_blue)[y * tlWidth + x];
+      if (color > 0) {
+        uint16_t c = colorize(color, logoColor);
+        canvas->drawPixel(larryStartX + x, larryStartY + y, c);
+      }
+    }
+  }
+}
+
+void animateText(Vec2d<int> &posVec, Vec2d<int> &posVisVec, Vec2d<int> &posHidVec)
+{
+  int dist = wakeupText ? (posVisVec.y - posHidVec.y) : (posHidVec.y - posVisVec.y);
+  int targetY = wakeupText ? posVisVec.y : posHidVec.y;
+
+  if (posVec.y != targetY) {
+    int mv = dist / totalTextAnimSteps;
+    posVec.y += mv;
+    if (mv > 0 && posVec.y > targetY) {
+      posVec.y = targetY;
+    } else if (mv <= 0 && posVec.y < targetY) {
+      posVec.y = targetY;
+    }
```

Comment:
this looks like it might be a linear interpolation could we use a lerp function to move these by the frame amount? i think the formula is something like 

function lerp(x, y, t) {
  return (1 - t) * x + t * y;
}

so t is the progress per frame, 1/totalAnimSteps maybe? usually a value like 0.1 or 0.2 is good too. then you can always feed in the current position as the x and the target as the y. the function will return the new position and the so on and so on. what do you think?

Reply:
Yeah, floats tend to give annoying rounding artifacts when working at small scales but I guess we can selectively ceil or floor based on travel direction to compensate.